modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java

```diff
@@ -326,14 +423,130 @@ private void increment() {
+
+        /** Is time logging enabled. */
+        private final boolean isTimeLoggingEnabled = IgniteSystemProperties.getBoolean(IGNITE_ENABLE_MESSAGES_TIME_LOGGING);
+
+        /** Metric bounds. */
+        private long[] metricBounds;
+
+        /**
+         * Reset metrics.
+         */
+        private void reset() {
+            outMetricsMap.clear();
+        }
+
+        /**
+         * Removes metrics data corresponding to {@code nodeId}
+         */
+        private void clearNodeMetrics(UUID nodeId) {
+            outMetricsMap.remove(nodeId);
+        }
+
+        /**
+         * Gets of obtains histogram metric bounds.
+         *
+         * @return Histogram metric bounds.
+         */
+        private long[] metricBounds() {
+            if (metricBounds == null)
+                metricBounds = obtainHistMetricBounds();
+
+            return metricBounds;
+        }
+
+        /**
+         * @return Metrics bounds in milliseconds.
+         */
+        private long[] obtainHistMetricBounds() {
+            String strBounds = IgniteSystemProperties.getString(IGNITE_COMM_SPI_TIME_HIST_BOUNDS);
+
+            if (strBounds == null)
+                return DEFAULT_HIST_BOUNDS;
+
+            try {
+                return Arrays.stream(strBounds.split(BOUNDS_PARAM_DELIMITER))
+                    .map(String::trim)
+                    .mapToLong(Long::parseLong)
+                    .toArray();
+            }
+            catch (Exception e) {
+                if (log != null)
+                    log.error("Wrong custom metric bounds format: " + strBounds + ". Using default: " +
+                              Arrays.toString(DEFAULT_HIST_BOUNDS));
+
+                return DEFAULT_HIST_BOUNDS;
+            }
+        }
+
+        /**
+         * Processes incoming messages.
+         *
+         * @param id Sender node id.
+         * @param msg Incoming message.
+         */
+        private void processIncomingMsg(UUID id, Message msg) {
+            if (!isTimeLoggingEnabled)
+                return;
+
+            if (msg instanceof TimeLoggableResponse)
+                addMetricsData(id, msg);
+            else if (msg instanceof TimeLoggableRequest) {
+                TimeLoggableRequest tlReq = (TimeLoggableRequest)msg;
+
+                tlReq.receiveTimestamp(System.nanoTime());
+            }
+        }
+
+        /**
+         * Adds data to hist metrics.
+         *
+         * @param nodeId Sender node id.
+         * @param msg Incoming message.
+         */
+        private void addMetricsData(UUID nodeId, Message msg) {
+            if (!(msg instanceof TimeLoggableResponse))
+                return;
+
+            TimeLoggableResponse timeLoggableRes = (TimeLoggableResponse)msg;
+
+            if (timeLoggableRes.responseSendTimestamp() == INVALID_TIMESTAMP)
+                return;
+
+            // No requests were sent to nodeId
+            outMetricsMap.putIfAbsent(nodeId, new ConcurrentHashMap<>());
```

Comment:
I think we should wrap `putIfAbsent` to condition that key isn't presented, for heap pressure reduction.